### PR TITLE
quiche: close the connection

### LINKED ISSUE #6213: curl + quiche: need to close the connection
## I did this

When I use curl (build with quiche) for http3 request, sometimes it stucks at n-th run, but it's random (60 - 300 range).

```
for i in $(seq 1 1000)
do
   curl --http3 https://http3-server/1k.png
done
```

### I expected the following

It should continuously run curl commands without failure.

When I look at server log, I have found the following:

- curl doesn't seems to properly close QUIC connection, by calling `quiche_conn_close()`
- this cause the server to close the connection by idle timeout
- before idle timeout expires, if next curl command try to reuse the same udp port which is not closed in the server, the server think it's a part of existing connection and drop the packet because it expects a short QUIC packet but long (Initial) packet is received. That's why it looks random (=up to the client OS picking up the source udp port)

### curl/libcurl version

```
$ curl -V
curl 7.74.0-DEV (x86_64-apple-darwin19.6.0) libcurl/7.74.0-DEV SecureTransport (BoringSSL) zlib/1.2.11 brotli/1.0.7 zstd/1.4.5 libidn2/2.3.0 nghttp2/1.41.0 quiche/0.6.0 librtmp/2.3
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli HTTP2 HTTP3 IDN IPv6 Largefile libz MultiSSL NTLM NTLM_WB SSL UnixSockets zstd
```

### operating system

```
$ uname -a
Darwin <hostname> 19.6.0 Darwin Kernel Version 19.6.0: Mon Aug 31 22:12:52 PDT 2020; root:xnu-6153.141.2~1/RELEASE_X86_64 x86_64
```

### PR DESCRIPTION
**Pull request by bagder**
Reported-by: Junho Choi (@junhochoi)
Fixes #6213

(I didn't find any docs for this function but copied the usage based on example code...)

### CODE REVIEW

File: lib/vquic/quiche.c [hunk 1]
```diff
@@ -100,6 +100,7 @@ static CURLcode qs_disconnect(struct quicsocket *qs)
     qs->cfg = NULL;
   }
   if(qs->conn) {
+    (void)quiche_conn_close(qs->conn, TRUE, 0, NULL, 0);
```
Comment (junhochoi) [inline lib/vquic/quiche.c:103]: I tested but the result is same. This only set the connection state to be closed, you need to run a usual event processing - see https://docs.rs/quiche/0.6.0/quiche/struct.Connection.html#method.close

Comment (bagder) [reply lib/vquic/quiche.c:103]: ah right, I forgot about that...

**Comment by bagder** [comment]
@junhochoi take 2: this now flushes the egress that at least has a chance of sending off outstanding packets. I'm not sure if this is good enough or if we actually have to also read packets etc.

**Comment by junhochoi** [comment]
> @junhochoi take 2: this now flushes the egress that at least has a chance of sending off outstanding packets. I'm not sure if this is good enough or if we actually have to also read packets etc.

Thanks! Now I can see CONNECTION_CLOSE is sent from curl at the end. Yes there will be a problem if this packet is lost, but since curl itself is terminating here, so I think this is good enough.

**Review by junhochoi — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**